token/core/tms.go

```diff
@@ -56,87 +56,151 @@ func NewTMSProvider(sp view.ServiceProvider, configProvider ConfigProvider, vaul
-	if len(network) == 0 {
+func (m *TMSProvider) GetTokenManagerService(opts driver.ServiceOptions) (service driver.TokenManagerService, err error) {
+	if len(opts.Network) == 0 {
 		return nil, errors.Errorf("network not specified")
 	}
-	if len(namespace) == 0 {
+	if len(opts.Namespace) == 0 {
 		return nil, errors.Errorf("namespace not specified")
 	}
 	m.lock.Lock()
-	defer m.lock.Unlock()
-
-	key := network + channel + namespace
-	service, ok := m.services[key]
-	if !ok {
-		if publicParamsFetcher == nil {
-			return nil, errors.Errorf("public params fetcher not specified")
+	invokeCallback := false
+	defer func() {
+		// unlock
+		m.lock.Unlock()
+		// invoke callback
+		if invokeCallback && m.callbackFunc != nil {
+			err = m.callbackFunc(service, opts.Network, opts.Channel, opts.Namespace)
+			if err != nil {
+				logger.Fatalf("failed to initialize tms for [%s]: [%s]", opts, err)
+			}
 		}
-		logger.Debugf("creating new token manager service for network %s, channel %s, namespace %s", network, channel, namespace)
+	}()
 
+	key := opts.Network + opts.Channel + opts.Namespace
+	var ok bool
+	service, ok = m.services[key]
+	if !ok {
+		logger.Debugf("creating new token manager service for [%s:%s:%s] with key [%s]", opts.Network, opts.Channel, opts.Namespace, key)
 		var err error
-		service, err = m.newTMS(network, channel, namespace, publicParamsFetcher)
+		service, err = m.newTMS(&opts)
 		if err != nil {
 			return nil, err
 		}
 		m.services[key] = service
+		invokeCallback = true
 	}
 	return service, nil
 }
 
-func (m *TMSProvider) newTMS(networkID string, channel string, namespace string, publicParamsFetcher driver.PublicParamsFetcher) (driver.TokenManagerService, error) {
-	driverName, err := m.driverFor(networkID, channel, namespace, publicParamsFetcher)
+func (m *TMSProvider) NewTokenManagerService(opts driver.ServiceOptions) (driver.TokenManagerService, error) {
+	if len(opts.Network) == 0 {
+		return nil, errors.Errorf("network not specified")
+	}
+	if len(opts.Namespace) == 0 {
+		return nil, errors.Errorf("namespace not specified")
+	}
+	logger.Debugf("creating new token manager service for [%s:%s:%s]", opts.Network, opts.Channel, opts.Namespace)
+
+	service, err := m.newTMS(&opts)
 	if err != nil {
-		return nil, errors.WithMessagef(err, "failed to get driver for [%s:%s:%s]", networkID, channel, namespace)
+		return nil, err
+	}
+	return service, nil
+}
+
+func (m *TMSProvider) Update(opts driver.ServiceOptions) error {
+	if len(opts.Network) == 0 {
+		return errors.Errorf("network not specified")
+	}
+	if len(opts.Namespace) == 0 {
+		return errors.Errorf("namespace not specified")
+	}
+	if len(opts.PublicParams) == 0 {
+		return errors.Errorf("public params not specified")
+	}
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	key := opts.Network + opts.Channel + opts.Namespace
```

Comment:
I would extract this into another method because this logic for the creation of the key is used on other places. Not to avoid code repetition but to be consistent on the construction of the key when we read and when we write.